File: packages/typescript-sdk/src/resources/calendar.ts
import { HTTPClient } from '../client/http';
import { Draft } from '../types';

export interface ContentPlan {
  id: string;
  user_id: string;
  days: DayPlan[];
  start_date: string;
  end_date: string;
  status: 'active' | 'completed' | 'archived';
  created_at: string;
}

export interface DayPlan {
  date: string;
  theme: string;
  content_ideas: ContentIdea[];
  posts: Draft[];
}

export interface ContentIdea {
  theme: string;
  caption_preview: string;
  platforms: string[];
  content_type: 'feed' | 'story' | 'carousel' | 'reel';
}

export interface GeneratePlanRequest {
  days: number;
  platforms: string[];
  themes?: string[];
  posting_frequency?: number; // posts per day
}

export interface TodayOverview {
  scheduled_posts: Draft[];
  draft_posts: Draft[];
  published_posts: Draft[];
  upcoming_key_dates: Array<{ date: string; label: string }>;
}

export class CalendarResource {
  constructor(private http: HTTPClient) {}

  /**
   * Get all content (drafts) for calendar view
   * @param startDate - Start date (YYYY-MM-DD)
   * @param endDate - End date (YYYY-MM-DD)
   */
  async getContent(startDate?: string, endDate?: string): Promise<{
    drafts: Draft[];
    total: number;
  }> {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);
    const query = params.toString() ? `?${params.toString()}` : '';

    return this.http.get<{ drafts: Draft[]; total: number }>(
      `/social-media/content-calendar${query}`
    );
  }

  /**
   * Get today's content overview
   */
  async getToday(): Promise<TodayOverview> {
    return this.http.get<TodayOverview>('/social-media/content-calendar/today');
  }

  /**
   * Get active content plan
   */
  async getPlan(): Promise<{ plan: ContentPlan | null }> {
    return this.http.get<{ plan: ContentPlan | null }>('/social-media/content-calendar/plan');
  }

  /**
   * Generate a new content plan
   * @param request - Plan generation parameters
   */
  async generatePlan(request: GeneratePlanRequest): Promise<{ plan: ContentPlan }> {
    return this.http.post<{ plan: ContentPlan }>(
      '/social-media/content-calendar/plan/generate',
      request
    );
  }

  /**
   * Regenerate content ideas for a specific day
   * @param planId - Content plan ID
   * @param dayIndex - Day index (0-based)
   */
  async regenerateDay(planId: string, dayIndex: number): Promise<{ day: DayPlan }> {
    return this.http.post<{ day: DayPlan }>(
      `/social-media/content-calendar/plan/${planId}/day/${dayIndex}/regenerate`
    );
  }

  /**
   * Create draft from content idea
   * @param planId - Content plan ID
   * @param dayIndex - Day index (0-based)
   * @param ideaIndex - Content idea index
   */
  async createDraftFromIdea(
    planId: string,
    dayIndex: number,
    ideaIndex: number
  ): Promise<{ draft: Draft }> {
    return this.http.post<{ draft: Draft }>(
      `/social-media/content-calendar/plan/${planId}/day/${dayIndex}/create-draft`,
      { idea_index: ideaIndex }
    );
  }
}

File: packages/python-sdk/src/urisocial/resources/calendar.py
"""Content calendar resource for URI Social SDK"""

from typing import Dict, Any, List, Optional
from ..http_client import HTTPClient


class CalendarResource:
    """Content calendar and planning"""

    def __init__(self, http: HTTPClient):
        self._http = http

    def get_content(
        self, start_date: Optional[str] = None, end_date: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get calendar content"""
        params = {}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        return self._http.get("/social-media/content-calendar", params=params)

    def get_today(self) -> Dict[str, Any]:
        """Get today's content overview"""
        return self._http.get("/social-media/content-calendar/today")

    def get_plan(self) -> Dict[str, Any]:
        """Get active content plan"""
        return self._http.get("/social-media/content-calendar/plan")

    def generate_plan(
        self,
        days: int,
        platforms: List[str],
        themes: Optional[List[str]] = None,
        posting_frequency: int = 1,
    ) -> Dict[str, Any]:
        """Generate a new content plan"""
        return self._http.post(
            "/social-media/content-calendar/plan/generate",
            json={
                "days": days,
                "platforms": platforms,
                "themes": themes,
                "posting_frequency": posting_frequency,
            },
        )

    def regenerate_day(self, plan_id: str, day_index: int) -> Dict[str, Any]:
        """Regenerate content ideas for a specific day"""
        return self._http.post(
            f"/social-media/content-calendar/plan/{plan_id}/day/{day_index}/regenerate"
        )

    def create_draft_from_idea(
        self, plan_id: str, day_index: int, idea_index: int
    ) -> Dict[str, Any]:
        """Create draft from content idea"""
        return self._http.post(
            f"/social-media/content-calendar/plan/{plan_id}/day/{day_index}/create-draft",
            json={"idea_index": idea_index},
        )
